File: app/src/main/java/com/ywwynm/everythingdone/database/DoingRecordDAO.java
package com.ywwynm.everythingdone.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.ywwynm.everythingdone.Def;
import com.ywwynm.everythingdone.model.DoingRecord;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ywwynm on 2016/11/9.
 * dao layer between model {@link DoingRecord} and table "doing_records".
 */
public class DoingRecordDAO {

    public static final String TAG = "DoingRecordDAO";

    private static DoingRecordDAO sDoingRecordDAO;

    public static DoingRecordDAO getInstance(Context context) {
        if (sDoingRecordDAO == null) {
            synchronized (DoingRecordDAO.class) {
                if (sDoingRecordDAO == null) {
                    sDoingRecordDAO = new DoingRecordDAO(context.getApplicationContext());
                }
            }
        }
        return sDoingRecordDAO;
    }

    private SQLiteDatabase db;

    private DoingRecordDAO(Context context) {
        DBHelper helper = new DBHelper(context);
        db = helper.getWritableDatabase();
    }

    public List<DoingRecord> getAllDoingRecords() {
        List<DoingRecord> reminders = new ArrayList<>();
        Cursor cursor = db.query(Def.Database.TABLE_DOING_RECORDS, null, null, null, null, null, null);
        while (cursor.moveToNext()) {
            reminders.add(new DoingRecord(cursor));
        }
        cursor.close();
        return reminders;
    }

    public boolean insert(DoingRecord doingRecord) {
        ContentValues values = new ContentValues();
        values.put(Def.Database.COLUMN_THING_ID_DOING,           doingRecord.getThingId());
        values.put(Def.Database.COLUMN_THING_TYPE_DOING,         doingRecord.getThingType());
        values.put(Def.Database.COLUMN_ADD5_TIMES_DOING,         doingRecord.getAdd5Times());
        values.put(Def.Database.COLUMN_PLAYED_TIMES_DOING,       doingRecord.getPlayedTimes());
        values.put(Def.Database.COLUMN_TOTAL_PLAY_TIME_DOING,    doingRecord.getTotalPlayTime());
        values.put(Def.Database.COLUMN_PREDICT_DOING_TIME_DOING, doingRecord.getPredictDoingTime());
        values.put(Def.Database.COLUMN_START_TIME_DOING,         doingRecord.getStartTime());
        values.put(Def.Database.COLUMN_END_TIME_DOING,           doingRecord.getEndTime());
        values.put(Def.Database.COLUMN_STOP_REASON_DOING,        doingRecord.getStopReason());
        values.put(Def.Database.COLUMN_START_TYPE_DOING,         doingRecord.getStartType());
        values.put(Def.Database.COLUMN_SHOULD_ASM_DOING,         doingRecord.shouldAutoStrictMode() ? 1 : 0);
        return db.insert(Def.Database.TABLE_DOING_RECORDS, null, values) != -1;
    }

}

File: app/src/main/java/com/ywwynm/everythingdone/database/DoingRecordDAO.kt
package com.ywwynm.everythingdone.database

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

import com.ywwynm.everythingdone.Def
import com.ywwynm.everythingdone.model.DoingRecord

import java.util.ArrayList

/**
 * Created by ywwynm on 2016/11/9.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * dao layer between model [DoingRecord] and table "doing_records".
 */
open class DoingRecordDAO private constructor(context: Context?) {

    private var db: SQLiteDatabase? = null

    init {
        val helper: DBHelper = DBHelper(context)
        db = helper.getWritableDatabase()
    }

    open fun getAllDoingRecords(): List<DoingRecord?>? {
        val reminders: MutableList<DoingRecord?> = ArrayList()
        val cursor: Cursor = db!!.query(Def.Database.TABLE_DOING_RECORDS, null, null, null, null, null, null)
        while (cursor.moveToNext()) {
            reminders.add(DoingRecord(cursor))
        }
        cursor.close()
        return reminders
    }

    open fun insert(doingRecord: DoingRecord?): Boolean {
        val values: ContentValues = ContentValues()
        values.put(Def.Database.COLUMN_THING_ID_DOING,           doingRecord!!.thingId)
        values.put(Def.Database.COLUMN_THING_TYPE_DOING,         doingRecord.thingType)
        values.put(Def.Database.COLUMN_ADD5_TIMES_DOING,         doingRecord.add5Times)
        values.put(Def.Database.COLUMN_PLAYED_TIMES_DOING,       doingRecord.playedTimes)
        values.put(Def.Database.COLUMN_TOTAL_PLAY_TIME_DOING,    doingRecord.totalPlayTime)
        values.put(Def.Database.COLUMN_PREDICT_DOING_TIME_DOING, doingRecord.predictDoingTime)
        values.put(Def.Database.COLUMN_START_TIME_DOING,         doingRecord.startTime)
        values.put(Def.Database.COLUMN_END_TIME_DOING,           doingRecord.endTime)
        values.put(Def.Database.COLUMN_STOP_REASON_DOING,        doingRecord.stopReason)
        values.put(Def.Database.COLUMN_START_TYPE_DOING,         doingRecord.startType)
        values.put(Def.Database.COLUMN_SHOULD_ASM_DOING,         if (doingRecord.shouldAutoStrictMode) 1 else 0)
        return db!!.insert(Def.Database.TABLE_DOING_RECORDS, null, values) != -1L
    }

    companion object {
        const val TAG: String = "DoingRecordDAO"

        @JvmField
        var sDoingRecordDAO: DoingRecordDAO? = null

        @JvmStatic
        fun getInstance(context: Context?): DoingRecordDAO? {
            if (sDoingRecordDAO == null) {
                synchronized(DoingRecordDAO::class.java) {
                    if (sDoingRecordDAO == null) {
                        sDoingRecordDAO = DoingRecordDAO(context!!.getApplicationContext())
                    }
                }
            }
            return sDoingRecordDAO
        }
    }
}
